nimbus/src/enrollment.rs

```diff
@@ -179,55 +525,117 @@ pub fn reset_enrollment(
         None => {
             // XXX - do we want specific errors for this kind of thing?
             log::warn!("No such experiment '{}'", experiment_slug);
-            return Ok(());
+            return Ok(vec![]);
         }
         Some(e) => e,
     };
-    let enrollment = evaluate_enrollment(nimbus_id, aru, app_context, &exp)?;
+    let merger = EnrollmentsMerger::new(nimbus_id, aru, app_context);
+    let is_user_participating = get_global_user_participation(db, writer)?;
+    let mut enrollment_events = vec![];
+    let enrollment = merger
+        .evolve_enrollment(
+            is_user_participating,
+            None, // Treat the experiment as we're seeing it for the first time...
+            Some(&exp),
+            None, // ...and we've never been enrolled to it.
+            &mut enrollment_events,
+        )?
+        .ok_or_else(|| {
+            Error::InternalError("reset_enrollment should have resulted in a non-empty enrollment")
+        })?;
     log::debug!(
         "Experiment '{}' reset enrollment {:?}",
         experiment_slug,
         enrollment
     );
     let enr_store = db.get_store(StoreId::Enrollments);
     enr_store.put(writer, experiment_slug, &enrollment)?;
-    Ok(())
+    Ok(enrollment_events)
 }
 
-pub fn opt_in_with_branch(db: &Database, experiment_slug: &str, branch: &str) -> Result<()> {
-    // For now we don't bother checking if the experiment or branch exist - if
-    // they don't the enrollment will just be removed next time we refresh.
+pub fn opt_in_with_branch(
+    db: &Database,
+    experiment_slug: &str,
+    branch: &str,
+) -> Result<Vec<EnrollmentChangeEvent>> {
+    let mut writer = db.write()?;
+    let exp_store = db.get_store(StoreId::Experiments);
+    let exp = match exp_store.get::<Experiment>(&writer, experiment_slug)? {
+        None => {
+            // XXX - do we want specific errors for this kind of thing?
+            log::warn!("No such experiment '{}'", experiment_slug);
+            return Ok(vec![]);
+        }
+        Some(e) => e,
+    };
+    if !exp
+        .branches
+        .iter()
+        .map(|b| &b.slug)
+        .any(|branch_slug| branch == branch_slug)
+    {
+        // XXX - do we want specific errors for this kind of thing?
+        log::warn!(
+            "No such branch '{}' for experiment {}",
+            branch,
+            experiment_slug
+        );
+        return Ok(vec![]);
+    }
     let enrollment = ExperimentEnrollment {
         slug: experiment_slug.to_string(),
-        status: EnrollmentStatus::Enrolled {
-            reason: EnrolledReason::OptIn,
-            branch: branch.to_string(),
-        },
+        status: EnrollmentStatus::enrolled(EnrolledReason::OptIn, branch),
+    };
+    let enrollment_id = if let EnrollmentStatus::Enrolled { enrollment_id, .. } = &enrollment.status
+    {
+        enrollment_id
+    } else {
+        unreachable!(); // This is really dumb.
```

Comment:
A `get_enrollment_id` helper method on `ExperimentEnrollment` might help clean this up a bit.